src/AggregateFunctions/AggregateFunctionWindowFunnel.cpp

```diff
@@ -169,55 +210,110 @@ class AggregateFunctionWindowFunnel final
 
         data.sort();
 
-        /// events_timestamp stores the timestamp of the first and previous i-th level event happen within time window
-        std::vector<std::optional<std::pair<UInt64, UInt64>>> events_timestamp(events_size);
-        bool first_event = false;
+        /// Stores the timestamp of the first and last i-th level event happen within time window
+        struct EventMatchTimeWindow
+        {
+            UInt64 first_timestamp;
+            UInt64 last_timestamp;
+            std::array<UInt64, MAX_EVENTS> event_path;
+
+            EventMatchTimeWindow() = default;
+            EventMatchTimeWindow(UInt64 first_ts, UInt64 last_ts)
+                : first_timestamp(first_ts), last_timestamp(last_ts) {}
+        };
+
+        /// We track all possible event sequences up to the current event.
+        /// It's required because one event can meet several conditions.
+        /// For example: for events 'start', 'a', 'b', 'a', 'end'.
+        /// The second occurrence of 'a' should be counted only once in one sequence.
+        /// However, we do not know in advance if the next event will be 'b' or 'end', so we try to keep both paths.
+        std::vector<std::list<EventMatchTimeWindow>> event_sequences(events_size);
+
+        bool has_first_event = false;
         for (size_t i = 0; i < data.events_list.size(); ++i)
         {
-            const T & timestamp = data.events_list[i].first;
-            const auto & event_idx = data.events_list[i].second - 1;
+            const auto & current_event = data.events_list[i];
+            auto timestamp = current_event.timestamp;
+            Int64 event_idx = current_event.event_type - 1;
+            UInt64 unique_id = current_event.unique_id;
+
             if (strict_order && event_idx == -1)
             {
-                if (first_event)
+                if (has_first_event)
                     break;
                 else
                     continue;
             }
             else if (event_idx == 0)
             {
-                events_timestamp[0] = std::make_pair(timestamp, timestamp);
-                first_event = true;
+                if (!strict_once)
+                    /// Do not keep all sequences, when mode is disabled
```

Comment:
Maybe implement separate method for `strict_once` that uses the sequence? It looks like it still gives overhead for other modes that affects [performance](https://s3.amazonaws.com/clickhouse-test-reports/69738/29ed3f4daf1d6cc0a5a80e6c848a1f1e9fb2697c/performance_comparison__release__[3_4]/report.html).